core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.persistence.cassandra.query
+
+import java.util.concurrent.ThreadLocalRandom
+
+import akka.annotation.InternalApi
+import akka.persistence.cassandra._
+import akka.stream.stage._
+import akka.stream.{Attributes, Outlet, SourceShape}
+import com.datastax.driver.core.{PreparedStatement, ResultSet, Session}
+
+import scala.collection.immutable.Queue
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContextExecutor
+
+/**
+  * INTERNAL API
+  */
+@InternalApi private[akka] object AllPersistenceIdsStage {
+
+  case object Continue
+
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi private[akka] final class AllPersistenceIdsStage(
+    refreshInterval: Option[FiniteDuration],
+    fetchSize: Int,
+    preparedStatement: PreparedStatement,
+    session: Session)
+    extends GraphStage[SourceShape[String]] {
+
+  import AllPersistenceIdsStage._
+
+  val out: Outlet[String] = Outlet("AllPersistenceIds.out")
+
+  val shape: SourceShape[String] = SourceShape(out)
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with StageLogging with OutHandler {
+
+      implicit def executionContext: ExecutionContextExecutor =
+        materializer.executionContext
+
+      private var queryInProgress = false
+      private var knownPersistenceIds: Set[String] = Set.empty
```

Comment:
@patriknw @chbatey @ennru what do you think about using immutables set and queue here? isn't it better to replace them with mutable versions? It is not public API anyway, it's internal private state. I firstly thought that immutable collections are more idiomatic and safe to use, but now I think it could affect performance. Do you have any performance tests for this query?

Reply:
I don't think we have any performance tests for this stage.

The operations we do on this set are 'add' and 'contains'. For those I don't see a reason to assume a mutable set would be faster: it might even be slower since it'd have to protect against concurrent access.

It does seem a bit wasteful to have all persistence id's in memory like this, it could be interesting to experiment with other implementations. That seems outside the scope of this PR though - the previous implementation also used an immutable `Set` so this PR shouldn't introduce a regression here.

Reply:
This query probably has much worse scalability problems, such as the actual Cassandra query. Let's not be too ambitious here, it's probably not used for anything "real".